### PR TITLE
Create a `TESTING.md` document

### LINKED ISSUE #44: Create a `TESTING.md` document
# User Story:

As an OSCAL team member, I need 

## Goals:

- A description of the testing methodology for this project, and how tests can be run
    - A summary of what is and is not tested
    - An explanation of how to run and add new tests
    - A metric defining the desired code coverage
- Conformity to SA-11 and (some) enhancements
    - SA-11(1): GitHub's static code analysis action
    - SA-11(2): Dependabot?
    - SA-11(4): CODEOWNERS
    - SA-11(7)
    - ?

## Dependencies:

This issue was formulated as a result of https://github.com/usnistgov/OSCAL/issues/1319

## Acceptance Criteria

- [ ] readme documentation affected by the changes in this issue have been updated.
- [ ] A Pull Request (PR) is submitted that fully addresses the goals of this User Story. This issue is referenced in the PR.
- [ ] The CI-CD build process runs without any reported errors on the PR. This can be confirmed by reviewing that all checks have passed in the PR.

## Note

- This issue is part of #43 
- A "follow-up" to this issue may be to OSCALize this document into a [component definition](https://pages.nist.gov/OSCAL/concepts/layer/implementation/component-definition/)

### PR DESCRIPTION
**Pull request by nikitawootten-nist**
Fixes #44

# Committer Notes

- Created a `TESTING.md` document as detailed in #44 
- Added a few key tests
- Added NYC coverage and made GH Actions use it instead of the test target

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/oscal-deep-diff/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/oscal-deep-diff/pulls) for the same update/change?
- [x] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Do all automated CI/CD checks pass?
- [x] Have you checked to see if any extraneous dependencies are living in [package.json](../package.json) or [package-lock.json](../package-lock.json)?

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [x] Have you included examples of how to use your new feature(s)?
- [x] Have you updated readme documentation affected by the changes you made?

### CODE REVIEW

**Review by wendellpiez — APPROVED** [review]
Sweet!

**Review by aj-stein-nist — COMMENTED** [review]
Sorry my comments here are long overdue. I have some ideas about structuring documentation where relevant around 800-53 Revision 5 SA controls and others around the assessment rubric in 800-53A and writing it a little more like controls to support the information security bent of the need for a model `TESTING.md` docs. Perhaps we talk more tomorrow about this?

**Comment by aj-stein-nist** [comment]
> Sorry my comments here are long overdue. I have some ideas about structuring documentation where relevant around 800-53 Revision 5 SA controls and others around the assessment rubric in 800-53A and writing it a little more like controls to support the information security bent of the need for a model `TESTING.md` docs. Perhaps we talk more tomorrow about this?

@nikitawootten-nist I will try and find some time for us to chat, but to be more clear, we should think about how we would address the specific criteria for assessment in tables like this screenshot below from [the PDF of 800-53A Revision 5](https://csrc.nist.gov/publications/detail/sp/800-53a/rev-5/final).

image

**Comment by aj-stein-nist** [comment]
@nikitawootten-nist, @david-waltermire-nist asked in our status update meeting if I want to discuss the currently open feedback (how to update content with an eye to 800-53A criteria), I should time box that to 1-2 hours with you and make sure it is good to go, barring no other feedback on the rest of the PR from others.

I will look for some time in our calendars soon. :-)

**Comment by aj-stein-nist** [comment]
@nikitawootten-nist I posted https://github.com/usnistgov/oscal-deep-diff/pull/45#discussion_r924730730 after we workshopped it. Let me know if you want to continue along by yourself or you would like me to pitch in on specific controls or control enhancements and specific statements, I can help out, just tell me which statements you want me to handle and you the rest. Lemme know!

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -1,75 +1,80 @@
...
+It provides basic information on the OSCAL project, describes the main ways people can make contributions, explains how to report issues with OSCAL, and lists pointers to additional sources of information.
 
 ## Project approach
 
 The approach we’re taking with OSCAL is agile. We’re adopting the philosophy of implementing the 20% of the functionality that solves 80% of the problem. We’re trying to focus on the core capabilities that are needed to provide the greatest amount of benefit. Because we’re working on a small set of capabilities, that allows us to make very fast progress. We’re building the features that we believe solve the biggest problems, so we’re providing the most value.
 
 We track our current work items using GitHub [project cards](https://github.com/usnistgov/OSCAL/projects). Our active project is typically the lowest numbered open project within the previously referenced page.
 
-*Note that the main OSCAL repository's projects are still used to track oscal-deep-diff progress through tracker tickets.*
+_Note that the main OSCAL repository's projects are still used to track oscal-deep-diff progress through tracker tickets._
```
Comment (aj-stein-nist) [inline CONTRIBUTING.md:12]: Non-blocking: still true?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -1,75 +1,80 @@
...
+This page is for potential contributors to the OSCAL project's oscal-deep-diff.
+It provides basic information on the OSCAL project, describes the main ways people can make contributions, explains how to report issues with OSCAL, and lists pointers to additional sources of information.
 
 ## Project approach
 
 The approach we’re taking with OSCAL is agile. We’re adopting the philosophy of implementing the 20% of the functionality that solves 80% of the problem. We’re trying to focus on the core capabilities that are needed to provide the greatest amount of benefit. Because we’re working on a small set of capabilities, that allows us to make very fast progress. We’re building the features that we believe solve the biggest problems, so we’re providing the most value.
 
 We track our current work items using GitHub [project cards](https://github.com/usnistgov/OSCAL/projects). Our active project is typically the lowest numbered open project within the previously referenced page.
```
Comment (aj-stein-nist) [inline CONTRIBUTING.md:10]: Minor and non-blocking, not sure we want to point to this now as the situation evolved as you patiently waited for my review. 😆 

```suggestion
We track our current work items using GitHub [project cards](https://github.com/orgs/usnistgov/projects/25/views/2). Our active project is typically the lowest numbered open project within the previously referenced page.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,75 +1,80 @@
...
 
 The core OSCAL project team regularly reviews the open issues, prioritizes their handling, and updates the issue statuses and comments as needed.
 
 ## Communications mechanisms
 
 There are two mailing lists for the project:
 
-- *oscal-dev@nist.gov* for communication among parties interested in contributing to the development of OSCAL or exchanging ideas. Subscribe by sending an email to [oscal-dev-join@nist.gov](mailto:oscal-dev-join@nist.gov). To unsubscribe send an email to [oscal-dev-leave@nist.gov](mailto:oscal-dev-leave@nist.gov).
-- *oscal-updates@nist.gov* for low-frequency updates on the status of the OSCAL project. Subscribe by sending an email to [oscal-updates-join@nist.gov](mailto:oscal-updates-join@nist.gov). To unsubscribe send an email to [oscal-updates-leave@nist.gov](mailto:oscal-updates-leave@nist.gov).
+-   *oscal-dev@nist.gov* for communication among parties interested in contributing to the development of OSCAL or exchanging ideas. Subscribe by sending an email to [oscal-dev-join@nist.gov](mailto:oscal-dev-join@nist.gov). To unsubscribe send an email to [oscal-dev-leave@nist.gov](mailto:oscal-dev-leave@nist.gov).
```
Comment (aj-stein-nist) [inline CONTRIBUTING.md:38]: Minor, [per our docs](https://pages.nist.gov/OSCAL/contact/#oscal-mailing-lists) and prior experience, the join/leave emails need updating, obviously minor.

```suggestion
-   *oscal-dev@nist.gov* for communication among parties interested in contributing to the development of OSCAL or exchanging ideas. Subscribe by sending an email to [oscal-dev+subscribe@list.nist.gov](mailto:oscal-dev+subscribe@list.nist.gov). To unsubscribe send an email to [oscal-dev+unsubscribe@list.nist.gov@nist.gov](mailto:oscal-dev+unsubscribe@list.nist.gov).
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,75 +1,80 @@
...
+-   *oscal-updates@nist.gov* for low-frequency updates on the status of the OSCAL project. Subscribe by sending an email to [oscal-updates-join@nist.gov](mailto:oscal-updates-join@nist.gov). To unsubscribe send an email to [oscal-updates-leave@nist.gov](mailto:oscal-updates-leave@nist.gov).
```
Comment (aj-stein-nist) [inline CONTRIBUTING.md:39]: Minor and non-blocking, ditto here.

```suggestion
-   *oscal-updates@nist.gov* for low-frequency updates on the status of the OSCAL project. Subscribe by sending an email to [oscal-updates+subscribe@list.nist.gov](mailto:oscal-updates+subscribe@list.nist.gov). To unsubscribe send an email to [oscal-updates+unsubscribe@list.nist.gov](mailto:oscal-updates+unsubscribe@list.nist.gov).
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,75 +1,80 @@
...
-- [src/](src): Houses the source code for this project.
-- [package.json](package.json): Configures the project npm package.
-- [package-lock.json](package-lock.json): Contains a snapshot of the npm dependencies for the project.
-- [tsconfig.json](tsconfig.json): Configures the typescript compiler settings for building the project.
-
+-   [.github/](.github): Contains GitHub issue and pull request templates for the OSCAL project.
+-   [.vscode/](.vscode): Contains editor-specific settings and run configurations.
+-   [bin/](bin): Contains node executable entry-points for oscal-deep-diff CLI.
+-   [examples/](examples): Contains example projects using this library.
+-   [meta/](meta): Contains mic. files
```
Comment (aj-stein-nist) [inline CONTRIBUTING.md:60]: Non-blocking, did you mean misc in place of mic.?

```suggestion
-   [meta/](meta): Contains miscellaneous files
```

File: TESTING.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+**SA-11: Developer Security Testing And Evaluation**:
+The NIST ITL CSD developers that maintain the `oscal-deep-diff` application system at all post-design stages of the system development life cycle:
+
+-   Perform unit and integration testing/evaluation for every commit in a development branch submitted for code review in the form of a pull request sent to the development team before merging it to the main release branch at the development team's recommended level of depth and coverage as described in the code coverage tool's configuration file [`.nycrc`](./.nyrc);
+-   Produce evidence of the execution of the assessment plan and the results of the testing and evaluation;
+-   Implement a verifiable flaw remediation process;
+-   Correct flaws identified during testing and evaluation
+    The required coverage is defined in this repository by the config file [`.nycrc`](./.nycrc).
+
+This check is performed by GitHub actions via the [Lint and Test workflow](./.github/workflows/test.yaml) for all pull requests.
```
Comment (aj-stein-nist) [inline TESTING.md:42]: Non-blocking: minor capitalization error.

```suggestion
This check is performed by GitHub Actions via the [Lint and Test workflow](./.github/workflows/test.yaml) for all pull requests.
```

**Review by aj-stein-nist — APPROVED** [review]
I found some minor typos and stuff, but excellent work. I know these are minor copy edit changes, but I think this is a great beachhead to eventually dogfood our own use of OSCAL to more clearly express and evidence our conformance with organizational regs, 800-53/A and otherwise!

File: CONTRIBUTING.md (hunk 1 above)
Comment (nikitawootten-nist) [reply CONTRIBUTING.md:12]: Good point, I'll update this line as well.

**Pull request merged**